### PR TITLE
Cancel order

### PR DESCRIPTION
**Pull request by TymofiiSavinov**

### CODE REVIEW

**Review by markblaize — CHANGES_REQUESTED** [review]

File: src/cancel_order.rs [hunk 1]
```diff
@@ -1,6 +1,193 @@
...
+            panic!("Order with id: {} not found", order_id.0);
+        });
+
+        //TODO: set real min_amount_x/min_amount_y
+        let amount = 1;
+        let min_amount_x = order.amount;
+        let min_amount_y = 0;
+
+        if order.status == OrderStatus::Pending {
+            ref_finance::ext(order.buy_token.clone())
```
[the remark this answers is not in the document]
Comment (TymofiiSavinov) [reply src/cancel_order.rs:93]: yep

File: src/cancel_order.rs [hunk 2]
```diff
@@ -1,6 +1,193 @@
...
+                        .with_attached_deposit(NO_DEPOSIT)
+                        .remove_liquidity_callback(order_id, swap_fee, price_impact),
+                );
+        } else {
+            self.remove_liquidity_callback(order_id, swap_fee, price_impact);
+        }
+    }
+
+    #[private]
+    pub fn remove_liquidity_callback(&self, order_id: U128, swap_fee: U128, price_impact: U128) {
```
[the remark this answers is not in the document]
Comment (TymofiiSavinov) [reply src/cancel_order.rs:114]: done

File: src/cancel_order.rs [hunk 3]
```diff
@@ -1,6 +1,193 @@
...
+            )
+            .then(
+                ext_self::ext(current_account_id())
+                    .with_static_gas(Gas(20))
+                    .with_attached_deposit(NO_DEPOSIT)
+                    .swap_callback(order_id, swap_fee, price_impact),
+            );
+    }
+
+    pub fn swap_callback(&mut self, order_id: U128, swap_fee: U128, price_impact: U128) {
```
[the remark this answers is not in the document]
Comment (TymofiiSavinov) [reply src/cancel_order.rs:160]: done

File: src/cancel_order.rs [hunk 4]
```diff
@@ -1,6 +1,198 @@
...
+                );
+        } else {
+            self.swap(order_id, swap_fee, price_impact);
+        }
+    }
+
+    #[private]
+    pub fn remove_liquidity_callback(&self, order_id: U128, swap_fee: U128, price_impact: U128) {
+        require!(is_promise_success(), "Some problem with remove liquidity");
+        self.swap(order_id, swap_fee, price_impact);
```
Comment (markblaize) [inline src/cancel_order.rs:63]: @TymofiiSavinov we actually should jump right to `cancel_swap_callback` here as we already have all USDT

File: src/cancel_order.rs [hunk 5]
```diff
@@ -1,6 +1,198 @@
...
+            .then(
+                ext_self::ext(current_account_id())
+                    .with_static_gas(Gas(20))
+                    .with_attached_deposit(NO_DEPOSIT)
+                    .cancel_swap_callback(order_id, swap_fee, price_impact),
+            );
+    }
+
+    #[private]
+    pub fn cancel_swap_callback(&mut self, order_id: U128, swap_fee: U128, price_impact: U128) {
```
Comment (markblaize) [inline src/cancel_order.rs:114]: @TymofiiSavinov why it's called `cancel_swap_callback` if the handler functions `swap`?

File: src/cancel_order.rs (hunk 4 above)
Comment (TymofiiSavinov) [reply src/cancel_order.rs]: we have swap_callback in create order, but with not same functionality. I changed name to order_cancel_swap_callback

File: src/cancel_order.rs (hunk 5 above)
Comment (TymofiiSavinov) [reply src/cancel_order.rs]: we have swap_callback in create order, but with not same functionality. I changed name to order_cancel_swap_callback

**Pull request merged**